src/lib/alloc.c

```diff
@@ -324,13 +324,11 @@ static struct mm_heap *get_heap_from_ptr(void *ptr)
 static struct mm_heap *get_heap_from_caps(struct mm_heap *heap, int count,
 					  uint32_t caps)
 {
-	uint32_t mask;
 	int i;
 
 	/* find first heap that support type */
 	for (i = 0; i < count; i++) {
-		mask = heap[i].caps & caps;
-		if (mask == caps)
+		if (heap[i].caps & caps)
```

Comment:
@tlauda 
Please image:
heap[i].caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_EXT
caps = SOF_MEM_CAPS_RAM |  SOF_MEM_CAPS_CACHE

Reply:
Yes, and in the case @libinyang shown the allocation for this heap will be rejected with current version because `heap[i].caps != caps` . This is wrong we should accept the request because both caps have common `SOF_MEM_CAPS_RAM` and this is exactly what this PR does :)

@tlauda , do you see it now?

Reply:
@libinyang Your example shows, that this change is even worse, because it actually breaks things. You are requesting memory which should have the RAM and CACHE capabilities. Why do you expect to have successful allocation on the heap, which supports only one of the caps?

Reply:
@tlauda, the question is how do we treat *CAPS*? When we pass 3 different caps along side with request do we want to allocate from first memory region which support either one of these caps OR we want to allocate from memory region which supports all of the caps?
I feel like only the former approach is correct - one of the caps is supported then we can accept request.

Reply:
> Maybe you just need more generic caps created instead of changing the current logic?

Well I think the reason why we send multiple CAPS in request is to have, as you said "generic caps". Meaning "please try to allocate from memory region which supports SOF_MEM_CAPS_RAM, but if you fail (not enough space?) instead of terminating the allocation please try with SOF_MEM_CAPS_HP". So this is the way I see it being use.

> What is your specific usage scenario? 

So we have an issue raised by @keyonjie https://github.com/thesofproject/sof/issues/1382. Basically, with complete, full WoV topology (multiple pipelines, many components, etc) one of the component (selector in this case but could be any other) failed the allocation of its `comp_buffer` because it requested `SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP | SOF_MEM_CAPS_CACHE | SOF_MEM_CAPS_DMA`. There is only one memory region which supports all of these caps, namely **heap hp buffer**. So it looks like this memory region was already consumed at the time selector tried to allocate its buffer. Since no other memory region has **exact same** mem caps whole allocation failed. If we change the logic to pic memory region which supports one of the requested caps, memory allocation continue with next block and and all works fine.

Reply:
Yes. I'm now agree @tlauda . Otherwise, how can we try to allocate the memory meet both CACHE AND DMA?

Reply:
@libinyang, with my patch you can not. However do we have such a use case to allocate memory region which supports both CACHE and DMA? Do we in general want to differentiate memory blocks in a way `This is for SOF_MEM_CAPS_CACHE AND SOF_MEM_CAPS_DMA` but not for `SOF_MEM_CAPS_CACHE` or `SOF_MEM_CAPS_DMA` separately?

Reply:
So I think it all boils down to the question **When we send multiple memory CAPS do me want to allocate memory from region which supports either one of the CAPS or only all of them?**